### ISSUE TITLE
ghcide not starting for stack projects with docker enabled builds

### ISSUE THREAD

**Issue by mariatsji** [digital-asset/ghcide#221]
Steps to reproduce:

Run a stack build with a stack.yaml containing
```
packages:
- .
docker:
  enable: true
resolver: lts-14.14
```

now running `ghcide` results in (I had to retry this command due to some docker-pull in the background that gave a different error message)

```
ghcide                                                                                                                            1
ghcide version: 0.0.4 (GHC: 8.6.5)
Ghcide setup tester in /Users/sjumilli/haskell/stackdocker.
Report bugs at https://github.com/digital-asset/ghcide/issues

Step 1/6: Finding files to test in /Users/sjumilli/haskell/stackdocker
Found 3 files

Step 2/6: Looking for hie.yaml files that control setup
Found 1 cradle

Step 3/6, Cradle 1/1: Implicit cradle for /Users/sjumilli/haskell/stackdocker
Cradle {cradleRootDir = "/Users/sjumilli/haskell/stackdocker", cradleOptsProg = CradleAction: stack}

Step 4/6, Cradle 1/1: Loading GHC Session
ghcide: (ExitFailure 1,"","")
CallStack (from HasCallStack):
  error, called at src/HIE/Bios/Cradle.hs:299:18 in hie-bios-0.2.1-CpNUPrWjqiyHklKCc1nhUy:HIE.Bios.Cradle```

**Comment by ndmitchell** [comment]
This seems an issue with hie-bios rather than ghcide, since it gets ExitFailure when it tries to load the session. CC @mpickering

**Comment by mpickering** [comment]
How does stack's docker integration work? Does it mean that `ghc` is not on the user path anymore?

**Comment by mariatsji** [comment]
Thank you for responding, guys.

I believe that with docker enabled stack builds, ghc is only available inside a spawned docker container, yes, and that one have to pass flags and params to ghc via stack commands, such as
`stack runghc -- --no-ghc-package-path X.hs`
or
`stack run --docker-run-args='-e ENV=dev'`

A small snippet from an [2015 fpcomplete-blog-post](https://www.fpcomplete.com/blog/2015/08/stack-docker) explaining the integration between stack and docker :

"When Docker is enabled in stack.yaml, every invocation of stack (with the exception of certain sub-commands) transparently re-invokes itself in an ephemeral Docker container which has the project root directory and the stack home (~/.stack) bind-mounted. The container exists only to provide the environment in which the build runs, nothing is actually written to the container's file-system (any writes happen in the bind-mounted directories) and it the container is destroyed immediately after stack exits (using docker run --rm)"

**Comment by mpickering** [comment]
Should be fixed once https://github.com/mpickering/hie-bios/pull/97 is merged.

**Comment by mpickering** [comment]
We looked into this a little (https://github.com/mpickering/hie-bios/pull/101) and came to the conclusion that it was out of scope for `hie-bios` because you really need to execute `ghcide` inside the container for anything to work. This is because all the installed packages and C libraries are only accessible from inside the container.


Therefore, you should use `stack exec ghcide` to run it inside the container and make sure `ghcide` is installed in the container.

**Comment by cocreature** [comment]
Thanks for looking into that. I agree that this can’t be fixed in `hie-bios`. Setting the executable to `stack exec ghcide` seems like the best option here.

**Comment by mariatsji** [comment]
That possibility never closed my mind, thanks!
